File: backend/src/services/freeDataClient.ts
/**
 * Free, no-signup web data sources.
 *
 * Used in place of Bright Data while that account is stuck behind payment
 * verification (see bright-data.md). Same call shape as the search/scrape
 * functions in brightdataClient.ts -- swapping back later is a one-line
 * change in webClient.ts, callers don't need to change.
 *
 * No anti-bot bypass here, so this will fail on heavily protected sites
 * (Google, LinkedIn, etc). DuckDuckGo's HTML endpoint and plain page fetches
 * are what's realistic without Bright Data.
 */

import * as cheerio from "cheerio";

const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " +
  "(KHTML, like Gecko) Chrome/124.0 Safari/537.36";

export type SearchResult = {
  title: string;
  url: string;
  snippet: string;
};

/** DuckDuckGo HTML search results -- no API key needed. */
export async function searchEngine(
  query: string,
  { maxResults = 8 }: { maxResults?: number } = {},
): Promise<SearchResult[]> {
  const response = await fetch("https://html.duckduckgo.com/html/", {
    method: "POST",
    headers: {
      "User-Agent": USER_AGENT,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams({ q: query }),
    signal: AbortSignal.timeout(20_000),
  });
  if (!response.ok) {
    throw new Error(`duckduckgo search failed: ${response.status} ${response.statusText}`);
  }

  const $ = cheerio.load(await response.text());
  const results: SearchResult[] = [];
  $(".result__body")
    .slice(0, maxResults)
    .each((_, element) => {
      const titleEl = $(element).find(".result__title a").first();
      if (titleEl.length === 0) return;
      results.push({
        title: titleEl.text().trim(),
        url: titleEl.attr("href") ?? "",
        snippet: $(element).find(".result__snippet").first().text().trim(),
      });
    });
  return results;
}

/** Fetch a URL and return readable text. Best-effort -- sites with bot
 * protection will simply fail here (that's the gap Bright Data closes). */
export async function scrapeAsMarkdown(
  url: string,
  { maxChars = 6000 }: { maxChars?: number } = {},
): Promise<string> {
  const response = await fetch(url, {
    headers: { "User-Agent": USER_AGENT },
    redirect: "follow",
    signal: AbortSignal.timeout(20_000),
  });
  if (!response.ok) {
    throw new Error(`fetch failed for ${url}: ${response.status} ${response.statusText}`);
  }

  const $ = cheerio.load(await response.text());
  $("script, style, nav, footer, header").remove();
  return $.root().text().split(/\s+/).join(" ").trim().slice(0, maxChars);
}

File: backend/app/services/free_data_client.py
"""Free, no-signup web data sources.

Used in place of Bright Data while that account is stuck behind payment
verification (see bright-data.md). Same call shape as the search/scrape
functions in brightdata_client.py -- swapping back later is a one-line
change in web_client.py, callers don't need to change.

No anti-bot bypass here, so this will fail on heavily protected sites
(Google, LinkedIn, etc). DuckDuckGo's HTML endpoint and plain page fetches
are what's realistic without Bright Data.
"""

import httpx
from bs4 import BeautifulSoup

USER_AGENT = (
    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 "
    "(KHTML, like Gecko) Chrome/124.0 Safari/537.36"
)


def search_engine(query: str, *, max_results: int = 8) -> list[dict]:
    """DuckDuckGo HTML search results -- no API key needed."""
    with httpx.Client(timeout=20, headers={"User-Agent": USER_AGENT}) as client:
        resp = client.post("https://html.duckduckgo.com/html/", data={"q": query})
        resp.raise_for_status()

    soup = BeautifulSoup(resp.text, "html.parser")
    results = []
    for result in soup.select(".result__body")[:max_results]:
        title_el = result.select_one(".result__title a")
        snippet_el = result.select_one(".result__snippet")
        if not title_el:
            continue
        results.append(
            {
                "title": title_el.get_text(strip=True),
                "url": title_el.get("href", ""),
                "snippet": snippet_el.get_text(strip=True) if snippet_el else "",
            }
        )
    return results


def scrape_as_markdown(url: str, *, max_chars: int = 6000) -> str:
    """Fetch a URL and return readable text. Best-effort -- sites with bot
    protection will simply fail here (that's the gap Bright Data closes)."""
    with httpx.Client(timeout=20, headers={"User-Agent": USER_AGENT}, follow_redirects=True) as client:
        resp = client.get(url)
        resp.raise_for_status()

    soup = BeautifulSoup(resp.text, "html.parser")
    for tag in soup(["script", "style", "nav", "footer", "header"]):
        tag.decompose()
    text = " ".join(soup.get_text(separator=" ").split())
    return text[:max_chars]
